src/java_bytecode/java_object_factory.cpp

```diff
@@ -28,20 +32,126 @@ Function: gen_nondet_init
+exprt gen_nondet_state::allocate_object(
+  const exprt& target_expr,
+  const typet& allocate_type,
+  const source_locationt &loc,
+  bool create_dynamic_objects)
+{
+  const typet& allocate_type_resolved=ns.follow(allocate_type);
+  const typet& target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  if(!create_dynamic_objects)
+  {
+    symbolt &aux_symbol=new_tmp_symbol(symbol_table);
+    aux_symbol.type=allocate_type;
+    aux_symbol.is_static_lifetime=true;
+
+    exprt object=aux_symbol.symbol_expr();
+    exprt aoe=address_of_exprt(object);
+    if(cast_needed)
+      aoe=typecast_exprt(aoe, target_expr.type());
+    code_assignt code(target_expr,aoe);
+    code.add_source_location()=loc;
+    init_code.copy_to_operands(code);
+    return aoe;
+  }
+  else
+  {
+    // build size expression
+    exprt object_size=size_of_expr(allocate_type, namespacet(symbol_table));
+
+    if(allocate_type.id()!=ID_empty && !object_size.is_nil())
+    {
+      // malloc expression
+      exprt malloc_expr = side_effect_exprt(ID_malloc);
+      malloc_expr.copy_to_operands(object_size);
+      typet result_type=pointer_typet(allocate_type);
+      malloc_expr.type()=result_type;
+      // Create a symbol for the malloc expression so we can initialise
+      // without having to do it potentially through a double-deref, which
+      // breaks the to-SSA phase.
+      symbolt &malloc_sym=new_tmp_symbol(symbol_table,"malloc_site");
+      malloc_sym.type=pointer_typet(allocate_type);
+      code_assignt assign = code_assignt(malloc_sym.symbol_expr(), malloc_expr);
+      code_assignt &malloc_assign = assign;
+      malloc_assign.add_source_location() = loc;
+      init_code.copy_to_operands(malloc_assign);
+      malloc_expr=malloc_sym.symbol_expr();
+      if(cast_needed)
+        malloc_expr=typecast_exprt(malloc_expr,target_expr.type());
+      code_assignt code(target_expr, malloc_expr);
+      code.add_source_location() = loc;
+      init_code.copy_to_operands(code);
+      return malloc_sym.symbol_expr();
+    }
+    else
+    {
+      // make null
+      null_pointer_exprt null_pointer_expr(to_pointer_type(target_expr.type()));
+      code_assignt code(target_expr, null_pointer_expr);
+      code.add_source_location() = loc;
+      init_code.copy_to_operands(code);
+      return exprt();
+    }
+  }
+}
+
+// Override type says to ignore the LHS' real type, and init it with the given
+// type regardless. Used at the moment for reference arrays, which are implemented
+// as void* arrays but should be init'd as their true type with appropriate casts.
+void gen_nondet_state::gen_nondet_init(
   const exprt &expr,
-  code_blockt &init_code,
-  const namespacet &ns,
-  std::set<irep_idt> &recursion_set,
   bool is_sub,
-  irep_idt class_identifier)
+  irep_idt class_identifier,
+  const source_locationt &loc,
+  bool skip_classid,
+  bool create_dynamic_objects,
+  const typet *override_type)
```

Comment:
this has a default value of `0`
```
    const typet *override_type=0);
```
which is actually used in the method to signal presence of a value. Would you prefer having a const reference and a Boolean value to signal type override?